artemis-commons/src/main/java/org/apache/activemq/artemis/utils/SizeAwareMetric.java

```diff
@@ -0,0 +1,240 @@
+   }
+
+   protected void over() {
+      if (overCallback != null) {
+         try {
+            overCallback.run();
+         } catch (Throwable e) {
+            logger.warn(e.getMessage(), e);
+         }
+      }
+   }
+
+   protected void under() {
+      if (underCallback != null) {
+         try {
+            underCallback.run();
+         } catch (Throwable e) {
+            logger.warn(e.getMessage(), e);
+         }
+      }
+   }
+
+   private boolean changeFlag(int expected, int newValue) {
+      return flagUpdater.compareAndSet(this, expected, newValue);
+   }
+
+   public long addSize(int delta) {
+      return addSize(delta, false);
+   }
+
+   public long addSizeOnly(int delta) {
+      return addSize(delta, true);
+   }
+
+   public long addSize(int delta, boolean sizeOnly) {
+
+      if (delta == 0) {
+         if (logger.isDebugEnabled()) {
+            logger.debug("SizeAwareMetric ignored with size 0", new Exception("trace"));
+         }
+         return sizeUpdater.get(this);
+      }
+
+      flagUpdater.compareAndSet(this, NOT_USED, FREE);
+
+      if (parentMetric != null) {
+         if (parentMetric != null) {
+            parentMetric.addSize(delta, sizeOnly);
+         }
+      }
+
+      IntConsumer theFunction = sizeOnly ? onSizeOnlyCallback : onSizeCallback;
+
+      if (theFunction != null) {
+         try {
+            theFunction.accept(delta);
+         } catch (Throwable e) {
+            logger.warn(e.getMessage(), e);
+         }
+      }
+
+      long currentSize = sizeUpdater.addAndGet(this, delta);
+
+      if (sizeOnly) {
+         long currentElements = elementsUpdater.get(this);
+         if (delta > 0) {
+            checkOver(currentElements, currentSize);
+         } else { // (delta < 0)
+            checkUnder(currentElements, currentSize);
+         }
+      } else {
+         if (delta > 0) {
+            checkOver(elementsUpdater.incrementAndGet(this), currentSize);
+         } else { // (delta < 0)
+            checkUnder(elementsUpdater.decrementAndGet(this), currentSize);
+         }
+      }
+
+      return currentSize;
+   }
+
+   public void setMax(long maxSize, long lowerMarkSize, long maxElements, long lowerMarkElements) {
+      this.maxSize = maxSize;
+      this.lowerMarkSize = lowerMarkSize;
+      this.maxElements = maxElements;
+      this.lowerMarkElements = lowerMarkElements;
+      long currentSize = sizeUpdater.get(this);
+      long currentElements = elementsUpdater.get(this);
+      checkOver(currentElements, currentSize);
+      checkUnder(currentElements, currentSize);
+   }
+
+   private void checkUnder(long currentElements, long currentSize) {
+      if (sizeEnabled) {
+         if (currentSize <= lowerMarkSize && changeFlag(OVER_SIZE, FREE)) {
+            under();
+         }
+      }
+
+      if (elementsEnabled) {
+         if (currentElements <= lowerMarkElements && changeFlag(OVER_ELEMENTS, FREE)) {
+            under();
+         }
+      }
+   }
```

Comment:
TLDR: Seems like it needs to account for the other value as well before simply firing 'under' whenever one of them goes under, or it might fire it incorrectly.

What if both go over but a later change only brings one of them under? Seems like it cant currently track that properly and so could actually fire the callback early and leave itself in the wrong state.

E.g whilst both can be over, the flag state only allows tracking that the size or the elements are over, so will only remember one of those (whichever flag is set first). Just because that remembered one goes under later, doesnt necessarily mean the other one is also under at that point, but currently the flag would just get reset straight to 'free' and the under callback would be fired, both possibly incorrectly.

Reply:
once it goes back from size, it can only go back from size... once it goes back from elements it can only go back from elements...  whatever hits it first.

once it goes back from SIZE, the flag will be OVER_SIZE and the other check will be moot. I don't see any need for any extra checks in there.

Reply:
@gemmellr fixed.. I amended the commit..

This is actually nice as it ensures the encapsulation here was a nice idea. fixing the check on an encapsulated class was better than doing the check on PagingStore and PagingManager.